### PR TITLE
Replace deprecated ReportIssue in more locations

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Follow up on #9422
Replaces 5 of the remaining 12 calls.

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs [hunk 1]
```diff
@@ -65,12 +65,12 @@ public sealed class CollectionsShouldImplementGenericInterface : SonarDiagnostic
...
                         {
-                            issues ??= new();
-                            issues.Add(Diagnostic.Create(Rule, typeDeclaration.Identifier.GetLocation(), suggestedGenericType));
+                            issues ??= [];
+                            issues.Add(new(typeDeclaration.Identifier, suggestedGenericType));
                         }
                     }
 
-                    issues?.ForEach(d => c.ReportIssue(d));
+                    issues?.ForEach(x => c.ReportIssue(Rule, x.Identifier, x.SuggestedGenericType));
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs:73]: This `ForEach` causes more damage than it solves. Let's refactor this a bit more:
* Remove `issues` variable and it's handling
* Remove `ForEach`
* Remove `Issue` record
* store the result of `containingType.Interfaces.Concat([containingType.BaseType]).WhereNotNull()` in an array.
* if all survive the current `return` condition, then `foreach` them. 
* inside foreach: if `SuggestGnericCollectionType` returns something, report directly

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotDecreaseMemberVisibility.cs [hunk 2]
```diff
@@ -41,14 +41,16 @@ public sealed class DoNotDecreaseMemberVisibility : SonarDiagnosticAnalyzer
...
+                    foreach (var issue in classDeclaration.Members.Select(issueFinder.FindIssue).WhereNotNull())
                     {
-                        c.ReportIssue(diagnostic);
+                        c.ReportIssue(Rule, issue.Location, issue.Symbol.ToDisplayString());
                     }
                 },
                 SyntaxKind.ClassDeclaration,
                 SyntaxKindEx.RecordClassDeclaration);
 
+        private record struct Issue(Location Location, ISymbol Symbol);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotDecreaseMemberVisibility.cs:52]: Similar to the previous: Change `IssueFinder` to `IssueReporter`, send him the context and let him report the issue directly.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotHideBaseClassMethods.cs [hunk 3]
```diff
@@ -41,14 +41,16 @@ public sealed class DoNotHideBaseClassMethods : SonarDiagnosticAnalyzer
...
+                    foreach (var issue in declarationSyntax.Members.Select(issueFinder.FindIssue).WhereNotNull())
                     {
-                        c.ReportIssue(diagnostic);
+                        c.ReportIssue(Rule, issue.Location, issue.HiddenBaseMethod.ToDisplayString());
                     }
                 },
                 SyntaxKind.ClassDeclaration,
                 SyntaxKindEx.RecordClassDeclaration);
 
+        private readonly record struct Issue(Location Location, IMethodSymbol HiddenBaseMethod);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotHideBaseClassMethods.cs:52]: Same here

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
These are bit more tricky

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs (hunk 1 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs]: Done in cec3919b05537bfb314a6a7555879011e0e6ecf9

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**